webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/services/servers/RemoveNodeService.scala

```diff
@@ -178,105 +233,63 @@ class RemoveNodeServiceImpl(
-            runPostHook   = RunHooks.syncRun(postHooks, hookEnv, systemEnv)
-            timePostHooks =  (System.currentTimeMillis - postHooksTime)
-            _             = logEffect.debug(s"Node-post-deletion scripts hooks ran in ${timePostHooks} ms")
-          } yield {
-            removeKeyCertification(nodeId).toBox match {
-              case Full(_) => //ok
-              case eb: EmptyBox =>
-                val e = (eb ?~! "Error when removing the certification status of node key")
-                logEffect.warn(e.messageChain)
-            }
-            // try to delete cfengine key if needed
-            deleteCfengineKey(nodeInfo)
-
-            runPostHook match {
-              case stop : HookReturnCode.Error => PostHookFailed(stop)
-              case _ => Success
-            }
-          }
-      }
-    }
-
-    logEffect.debug("Trying to remove node %s from the LDAP".format(nodeId.value))
-    (nodeId.value match {
-      case "root" => Failure("The root node cannot be deleted from the nodes list.")
+  override def removeNodePure(nodeId: NodeId, mode: DeleteMode, modId: ModificationId, actor: EventActor) : IOResult[NodeInfo] = {
+    // main logic, see help function below
+    nodeId match {
+      case Constants.ROOT_POLICY_SERVER_ID => Inconsistency("The root node cannot be deleted.").fail
       case _ => {
-
         for {
-          optNodeInfo   <- nodeInfoService.getNodeInfo(nodeId)
-
-          nodeInfo      <- optNodeInfo match {
-                             case None    => Failure(s"The node with id ${nodeId.value} was not found and can not be deleted")
-                             case Some(x) => Full(x)
-                           }
-          nodePaths     =  getNodePath(nodeInfo) match {
-                             case Full(x) => Some(x)
-                             case eb:EmptyBox => // if the policy server is not found, we must not fails (#11231)
-                               val msg = (eb ?~! s"Error when trying to calculate node '${nodeId.value}' policy path").messageChain
-                               logEffect.warn(msg)
-                               None
-                           }
-          startPreHooks =  System.currentTimeMillis
-          preHooks      <- RunHooks.getHooks(HOOKS_D + "/node-pre-deletion", HOOKS_IGNORE_SUFFIXES)
-          result        <- effectiveDeletion(nodeInfo, nodePaths, preHooks, startPreHooks)
+          _      <- NodeLoggerPure.Delete.debug(s"Deleting node with ID '${nodeId.value}' [mode:${mode.name}]")
+          status <- findNodeStatus(nodeId)
+          -      <- NodeLoggerPure.Delete.debug(s"  - node '${nodeId.value}' has status: [${status.map(_.name).mkString(",")}]")
+          info   <- Ref.make(Option.empty[NodeInfo]) //a place to store the maybe node info
+          // always delete in order pending then accepted then deleted
+           res1  <- if(status.contains(PendingInventory)) {
```

Comment:
Just a question, why this order ? and why pending first and not maybe Removed ?  I can understand why Accepted later since it has much more impact (and may be the last one imo

Reply:
Although i would have fold/accumulate here, instead of doing it later with the merge function

Reply:
I wanted to be sure that "ou=Removed Inventories" was cleaned in all cases, even we other parts of rudder were moving things to it

Reply:
Makes sense!